File: BinaryArraySearch/binary_search.py
from time import time


def contain(collection, target):
    """Determine whether list contains target"""
    return target in collection


def bs_contains(orderd, target):
    """
    Binary search for determine ordered list contains target
    :return position if target in collection
    """
    low = 0
    high = len(orderd) - 1

    while low <= high:
        mid = (low + high)//2
        if target == orderd[mid]:
            return mid
        elif target < orderd[mid]:
            high = mid - 1
        else:
            low = mid + 1
    return -(low + 1)


def insert_in_place_simple(ordered, target):
    """Insert target in collection"""
    for i in range(len(ordered)):
        if target < ordered[i]:
            ordered.insert(i, target)
            return
    ordered.append(target)


def insert_in_place_bs(ordered, target):
    """Insert target in collection"""
    idx = bs_contains(ordered, target)
    if idx < 0:
        ordered.insert(-(idx + 1), target)
        return
    ordered.insert(idx, target)


def check_performance(function):
    """Check performace of search method"""
    n = 1024
    while n < 999999:
        sorted_list = list(range(n))

        now = time()
        function(sorted_list, n/4)
        done = time()

        print('{0} {1}'.format(n, (done-now)*1000))

        n *= 2


if __name__ == '__main__':
    check_performance(insert_in_place_simple)
    print ('=====================================')
    check_performance(insert_in_place_bs)
    print ('=====================================')
    check_performance(contain)
    print ('=====================================')
    check_performance(bs_contains)

File: BinaryArraySearch/binary_search.go
package main

import (
	"fmt"
	"sort"
	"time"
)

type checkFunction func([]int, int) int

func generateSlice(n int) []int {
	var slice []int
	for i := 0; i <= n; i++ {
		slice = append(slice, i)
	}
	return slice
}

// Simple realization of search element
func contain(collection []int, target int) int {
	for idx, value := range collection {
		if value == target {
			return idx
		}
	}
	return -1
}

// Binary search for determine ordered list contains target and return number of element
func bsContain(collection []int, target int) int {
	var mid int
	low := 0
	high := len(collection) - 1
	for low <= high {
		mid = (low + high) / 2
		if target == collection[mid] {
			return mid
		} else if target < collection[mid] {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}
	return -(low + 1)
}

func checkPerformance(function checkFunction) {
	n := 1024
	for n < 99999999 {
		sorted_list := generateSlice(n)

		now := time.Now()
		function(sorted_list, n/4)
		elapsed := time.Since(now)

		fmt.Println(n, elapsed)

		n *= 2
	}
}

func main() {
	checkPerformance(contain)
	fmt.Println("================")
	checkPerformance(bsContain)
	fmt.Println("================")
	checkPerformance(sort.SearchInts)

}
